### ISSUE TITLE
Enum: writing to db fails, when global-option is automatic or db-field is primary-key

### ISSUE THREAD

**Issue by Garios** [mikependon/RepoDb#1100]
#### Problem
I have a table with a primary-key ID [int]. The corresponding csharp class has a field ID, which is an enum [EnumTest:int]. If I make an insert on this table, then I get the exception "Compiler.Entity/Object.Property: Failed to automatically convert the value expression..."

The problem seems to be in Compiler.cs. In cases of Global-Option "Automatic" or when db-field is a primary-key it uses the "auto-conversion-handling", which doesn't cover enums. In this code-area it throws the exception. The code-block after would handle the enum-conversion, but to late:

**Origin Code: Compiler.cs**
```csharp
internal static Expression GetEntityInstancePropertyValueExpression(Expression entityInstanceExpression, ClassProperty classProperty, DbField dbField)
{
	var expression = (Expression)Expression.Property(entityInstanceExpression, classProperty.PropertyInfo);

	// Target type
	var handlerInstance = classProperty.GetPropertyHandler() ?? PropertyHandlerCache.Get<object>(dbField.Type.GetUnderlyingType());
	var targetType = GetPropertyHandlerSetParameter(handlerInstance)?.ParameterType ?? dbField.Type;

	// Auto-conversion Handling
	if (GlobalConfiguration.Options.ConversionType == ConversionType.Automatic || dbField?.IsPrimary == true || dbField?.IsIdentity == true)
	{
		try
		{
//THIS LINE FAILS
			expression = ConvertExpressionWithAutomaticConversion(expression, targetType?.GetUnderlyingType());
		}
		catch (Exception ex)
		{
			throw new InvalidOperationException($"Compiler.Entity/Object.Property: Failed to automatically convert the value expression for " +
				$"property '{classProperty.GetMappedName()} ({classProperty.PropertyInfo.PropertyType.FullName})'. {classProperty}", ex);
		}
	}

	/*
		* Note: The other data provider can coerce the Enum into its destination data type in the DB by default,
		*       except for PostgreSQL. The code written below is only to address the issue for this specific provider.
		*/

	// Enum Handling
	if (classProperty.PropertyInfo.PropertyType.GetUnderlyingType().IsEnum == true)
	{
		try
		{
			if (!IsPostgreSqlUserDefined(dbField))
			{
				var dbType = classProperty.GetDbType() ?? classProperty.PropertyInfo.PropertyType.GetUnderlyingType().GetDbType();
				var toType = dbType.HasValue ? new DbTypeToClientTypeResolver().Resolve(dbType.Value) : targetType?.GetUnderlyingType();
				expression = ConvertEnumExpressionToTypeExpression(expression, toType);
			}
		}
		catch (Exception ex)
		{
			throw new InvalidOperationException($"Compiler.Entity/Object.Property: Failed to convert the value expression from " +
				$"enumeration '{classProperty.PropertyInfo.PropertyType.FullName}' to type '{targetType?.GetUnderlyingType()}'. {classProperty}", ex);
		}
	}

	// Property Handler
	try
	{
		expression = ConvertExpressionToPropertyHandlerSetExpression(
			expression, null, classProperty, dbField?.Type.GetUnderlyingType());
	}
	catch (Exception ex)
	{
		throw new InvalidOperationException($"Compiler.Entity/Object.Property: Failed to convert the value expression for property handler '{handlerInstance?.GetType()}'. " +
			$"{classProperty}", ex);
	}

	// Return the Value
	return ConvertExpressionToTypeExpression(expression, StaticType.Object);
}
```

**Possible fix: Compiler.cs**
I think it's better to check for enum before doing auto-conversion.

```csharp
internal static Expression GetEntityInstancePropertyValueExpression(Expression entityInstanceExpression, ClassProperty classProperty, DbField dbField)
{
	var expression = (Expression)Expression.Property(entityInstanceExpression, classProperty.PropertyInfo);

	// Target type
	var handlerInstance = classProperty.GetPropertyHandler() ?? PropertyHandlerCache.Get<object>(dbField.Type.GetUnderlyingType());
	var targetType = GetPropertyHandlerSetParameter(handlerInstance)?.ParameterType ?? dbField.Type;


	/*
		* Note: The other data provider can coerce the Enum into its destination data type in the DB by default,
		*       except for PostgreSQL. The code written below is only to address the issue for this specific provider.
		*/

	// Enum Handling
	if (classProperty.PropertyInfo.PropertyType.GetUnderlyingType().IsEnum == true &&  !IsPostgreSqlUserDefined(dbField))
	{
		try
		{
				var dbType = classProperty.GetDbType() ?? classProperty.PropertyInfo.PropertyType.GetUnderlyingType().GetDbType();
				var toType = dbType.HasValue ? new DbTypeToClientTypeResolver().Resolve(dbType.Value) : targetType?.GetUnderlyingType();
				expression = ConvertEnumExpressionToTypeExpression(expression, toType);
		}
		catch (Exception ex)
		{
			throw new InvalidOperationException($"Compiler.Entity/Object.Property: Failed to convert the value expression from " +
				$"enumeration '{classProperty.PropertyInfo.PropertyType.FullName}' to type '{targetType?.GetUnderlyingType()}'. {classProperty}", ex);
		}
	}
	else if (GlobalConfiguration.Options.ConversionType == ConversionType.Automatic || dbField?.IsPrimary == true || dbField?.IsIdentity == true)
	{
		// Auto-conversion Handling
		try
		{
			expression = ConvertExpressionWithAutomaticConversion(expression, targetType?.GetUnderlyingType());
		}
		catch (Exception ex)
		{
			throw new InvalidOperationException($"Compiler.Entity/Object.Property: Failed to automatically convert the value expression for " +
				$"property '{classProperty.GetMappedName()} ({classProperty.PropertyInfo.PropertyType.FullName})'. {classProperty}", ex);
		}
	}

	// Property Handler
	try
	{
		expression = ConvertExpressionToPropertyHandlerSetExpression(
			expression, null, classProperty, dbField?.Type.GetUnderlyingType());
	}
	catch (Exception ex)
	{
		throw new InvalidOperationException($"Compiler.Entity/Object.Property: Failed to convert the value expression for property handler '{handlerInstance?.GetType()}'. " +
			$"{classProperty}", ex);
	}

	// Return the Value
	return ConvertExpressionToTypeExpression(expression, StaticType.Object);
}
```

**Comment by mikependon** [comment]
Wierd! Pretty sure this has been working in the past, but someway along the road, this bug has been introduced and is not captured by the Integration Test. FYI: As of writing this, this error is also happening even on a simple Automatic conversion use-case of enum.

**Comment by mikependon** [comment]
The fixes to this issue is available on the next version > RepoDB v1.13.0-alpha2.
